contracts/token-vesting.clar

```diff
@@ -5,12 +5,48 @@
 ;; Withdraws are unlocked to a specified set of addresses
 ;; at a certain schedule and conditions.
 
-;; Treasury map. Stores a "depositor" => "vesting" relation.
+;; SIP-010 token trait.
+(use-trait ft-trait .sip-010-trait.sip-010-trait)
+
+;;
+;; -- Data
+;;
+
+;; Treasury map.
+;; Identify and stores vestings for the contract.
 (define-map treasury
-  principal {token: (optional principal), amount: uint})
+  uint {sender: principal, amount: uint, token-contract: principal})
+
+;; Vesting ID.
+(define-data-var vesting-id uint u0)
+
+;; Vesting Vault.
+(define-data-var vesting-vault principal 'ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5)
+
+;;
+;; -- Public
+;;
 
 ;; Deposit function.
-(define-public (deposit (token principal) (amount uint))
+(define-public (deposit (token <ft-trait>) (amount uint))
   (begin
-    (map-set treasury tx-sender {token: (some token), amount: amount})
+    (map-set treasury
+      (get-next-vesting-id)
```

Comment:
Q: why is the ID incremented on each deposit? what's the logic behind this?

Reply:
@philipdesmedt I was thinking in the "reusability" of the contract. For instance, a person might want to use the contract to lock different SIP-010 tokens. Thus, each `id` would be a new vesting from the same contract owner. What you think about that use case?

Reply:
@csgui Yeah, that makes sense.. I'm not sure I would use an ID as key though.

Shouldn't we use the principal of a SIP10 token as key? Technically each token has only 1 treasury entry and has a vesting on it, so maybe the data structure could look something like this

map of a compound key: principal (e.g. using (contract-of token),`'SP2C2YFP12AJZB4MABJBAJ55XECVS7E4PMMZ89YZR.arkadiko-token`) and principal `tx-sender`.. so for each address + token combination you know which amount vests when. I would save the values as `token-amount`, `amount-per-vest`, `start-block-height`, `end-block-height`, where amount per vest is the amount of tokens unlocked per vesting period, the start block height indicates when the first amount of tokens vest (so you can easily introduce a cliff), end block height when the last amount of tokens vest. Maybe you need some additional maps then to make lookups easier.

Just thinking out loud here.. there could be other viable data structures, but this seems the easiest to work with in my mind. I could be wrong

Reply:
So something like this?

```
(define-map vesting
  { contract: principal, recipient: principal }
  {
    token-amount: uint,
    amount-per-period: uint,
    start-block-height: uint,
    end-block-height: uint
  }
)
(define-map vestings-per-principal
  { recipient: principal }
  { contracts: (list 100 principal) }
)
```

Lists are not super performant and it limits the amount of vestings per token a user can have, but we could replace that with another compound key with the ID in it and iterate that way. Again, there might be multiple good solutions